_posts/2019-02-05-configurable-builds-part-1.md

```diff
@@ -0,0 +1,280 @@
+targeting? What features do you want to include? What if you want to build a
+complex app with generated sources, client and server modules, native
+extensions, and test data?
+
+Bazel's historical approach of ad hoc flags:
+
+    $ bazel build //my:binary --cpu=arm --crosstool_top=//my:custom_toolchain --define MYFEATURE=1
+
+is lacking in many ways:
+
+1. `--cpu` only accepts values explicitly supported by C++ rules. Even if you're
+   not building C++.
+1. `--crosstool_top`, which specifies exactly what command line should compile
+   your code, is even more tied to C++. Other languages may be organized
+   completely differently.
+1. `--define` is completely unstructured and unreadable in Starlark.
+1. `$ bazel build //my:all` sets the same flags for all targets, even if
+   different targets need different flags.
+1. `//my:binary`'s deps must use the same settings as their parent.
+1. It might be hard to remember which flags are needed by which targets,
+   especially across users and dev/test/prod environments. This risks
+   inconsistent and possibly incorrect builds.
+
+A different approach might be to set flags directly in targets that need them,
+let rule writers design the flags that affect their rules, and provide standard
+extensible APIs for cross-language concepts like *platform* and *cpu*. This
+lets people configure their projects on their own terms, using language
+consistent with how their projects are organized.
+
+Bazel's configurability work is redesigning Bazel around these ideas. It's
+ultimate goal is to make
+
+    $ bazel build //:all
+
+*just work*, which means every target knows how to build "the right way", for
+whatever platform(s) and features you care about, and you only have to set
+flags that are genuinely interesting to you. All while keeping builds fast and
+correct.
+
+## Configuration vs. Attributes
+
+Rules have always had [attributes](), which also affect how they build. The
+difference between configuration and attributes is that attributes only affect
+the rule's [direct build actions](https://docs.bazel.build/versions/master/skylark/rules.html#implementation-function)).
+This means attributes **cannot affect a rule's dependencies**.
+
+For example, C++ rules have an attribute named
+[`copts`](https://docs.bazel.build/versions/master/be/c-cpp.html#cc_binary.copts)
+that sets custom C++ compile options. `cc_binary(name = "mybinary", srcs =
+["mybinary.cc"], deps = [":mylibrary"], copts =
+["-DUSE_EXPERIMENTAL_FEATURES=1"])` might compile `mybinary.cc` with
+experimental features. But this won't happen for `:mylibrary` unless it sets
+its `copts` similarly. This makes attributes unsuitable for tasks like building
+a binary for a different CPU.
+
+If you imagine a build as a [graph](https://docs.bazel.build/versions/master/build-ref.html#dependencies)
+with parents above their dependencies, attributes change behavior
+[*up*](https://docs.bazel.build/versions/master/skylark/rules.html#providers)
```

Comment:
Is providers the best link here? The link is not immediately obvious to someone who has not written a rule yet.

Reply:
I can't find any better reference. The key here is that providers are *the* interface we use for propagating info up the build graph. So it seems like an important connection.

Reply:
Closing. Feel free to re-open as desired.